### DISCUSSION TITLE
VL53L1X (ToF) sensor doesn't update values in the console with firmware tasmota-4M

### DISCUSSION THREAD

**Discussion by jackpotcityco** [arendst/Tasmota#19764] [Q&A] [answered]
Hello,

**Current Firmware:**
```
Program Version	        13.1.0.4(tasmota-4M)
Build Date & Time	2023-10-15T12:52:50

ESP Chip Id	        1574764 (ESP8266EX)
Flash Chip Id	        0x16405E (DOUT)
Flash Size	                 4096 KB
Program Flash Size	         4096 KB
Program Size	                  708 KB
Free Program Space	         1336 KB
Free Memory	                 22.5 KB
File System Size	         2024 KB
```

I have just installed the firmware: `tasmota-4M` which support `Conditional Rules (IF, ELSEIF, ELSE)`
The Process I used: https://github.com/arendst/Tasmota/discussions/19747#discussioncomment-7287486

The problem that happened after I installed this firmware is that the `VL53L1X - Time-of-Flight (ToF) ranging sensor`
doesn't update distance values in the Web Interface console as it did before I installed the new Firmware.
The firmware I had before which updated sensor values was: `Program Version 13.1.0.4(tasmota)`

It is my first time I install a firmware, so perheps I could have missed something?

**Below is the settings I have done for the new firmware:** `tasmota-4M`

- As before I have the `SDA cable` from the sensor connected to pin: `D2` on the `ESP8266EX`
- As before I have the `SCL cable` from the sensor connected to pin: `D1` on the `ESP8266EX`
- `GND` and `3V3` are also connected as before
![image](https://github.com/arendst/Tasmota/assets/47510843/0fd642b6-ef55-4cf3-ac1a-fd81c436028a)

**I have entered this Rule1 in the console to update** `VL53L1X sensor values`
`Rule1 ON VL53L1X#Distance DO var1 %value% ENDON`

**Below is the console commands:**
![image](https://github.com/arendst/Tasmota/assets/47510843/84b98420-3b6d-4d49-ad0e-21e72a0a1480)

Thank you!

**Comment by barbudor** [comment, marked as answer]
tasmota-4M doesn't have the driver for the VL52L1X => https://tasmota.github.io/docs/BUILDS/#available-features-and-sensors
Using `tasmota-4M` was just a way to get IF/ELSE quickly, avoid to compile your own binary

Jumping from one discussion to another makes it difficult to follow what you want to achieve
If you want to have both VL52L1X and the IF/ELSE, then there is no other way than compiling your own binary as there is no standard build that have both

https://tasmota.github.io/docs/Compile-your-build/

Choose the method that suits you

You need now to compile your own `tasmota`  or ` tasmota-4M` binary using the following in your `user_config_override.h`:
```C++
#ifndef USE_VL53L1X
#define USE_VL53L1X
#endif

#ifndef USE_EXPRESSION
#define USE_EXPRESSION
#endif

#ifndef SUPPORT_IF_STATEMENT
#define SUPPORT_IF_STATEMENT
#endif
```

**Reply by sfromis** [reply]
Anyway, there are no precompiled binaries even including `USE_VL53L1X` alone.

**Comment by jackpotcityco** [comment]
> You need now to compile your own tasmota or  tasmota-4M binary using the following in your user_config_override.h

Yes, I understand now. I would simply need to compile a custom firmware that includes what is needed like you showed there.

> Jumping from one discussion to another makes it difficult to follow what you want to achieve

You are right, what I am trying to acheive as a first step is to be able to use `Rules1-3` where I can use `IF, ELSEIF, ELSE` when writing Rules in the console. At this moment I only have 1 sensor which is the `VL53L1X` to start my learning with. 
I will try to go back and see how far I can manage to compile a custom firmware that includes the `USE_VL53L1X` and `SUPPORT_IF_STATEMENT`. It will be very useful to be able to do that as more specific sensors would be included later on.

_(I could not find a button to 'mark as answer')_ 
Thank you for your help!
